### PR TITLE
Nexus service store Cassandra implementation

### PR DESCRIPTION
**Pull request by pdoerner**
**What changed?**
Builds on https://github.com/temporalio/temporal/pull/5128
Adds Cassandra implementation for nexus service persistence

**Why?**
Need to persist nexus service info in Cassandra

**How did you test it?**

**Potential risks**
None

**Is hotfix candidate?**
No

### CODE REVIEW

File: common/persistence/persistence_interface.go [hunk 1]
```diff
@@ -740,23 +740,32 @@ type (
...
 
-	// InternalGetNexusIncomingServiceResponse is the response to GetNexusIncomingService
-	InternalGetNexusIncomingServiceResponse struct {
+	// InternalNexusIncomingService is the internal representation of an incoming Nexus service
+	InternalNexusIncomingService struct {
+		ServiceName   string
+		NamespaceID   string
+		TaskQueueName string
+		Metadata      *commonpb.DataBlob
+		Version       int64
```
[the remark this answers is not in the document]
Comment (bergundy) [reply common/persistence/persistence_interface.go:749]: You don't need this now that we're using service IDs.

File: common/persistence/cassandra/nexus_service_store.go [hunk 2]
```diff
@@ -61,30 +93,172 @@ func (s *NexusServiceStore) Close() {
...
 ) error {
-	return serviceerror.NewUnimplemented("Cassandra DeleteNexusIncomingService is not implemented")
+	query := s.session.Query(templateDeleteIncomingServiceQuery,
+		constServicesPartition,
+		rowTypeIncomingNexusService,
+		serviceID,
+	).WithContext(ctx)
+
+	previous := make(map[string]interface{})
+	_, err := query.MapScanCAS(previous)
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:258]: Wondering why you need CAS here, is it just to know that the delete was applied?

File: common/persistence/cassandra/nexus_service_store.go [hunk 3]
```diff
@@ -61,30 +93,172 @@ func (s *NexusServiceStore) Close() {
 	}
 }
 
+func (s *NexusServiceStore) CreateOrUpdateNexusIncomingService(
+	ctx context.Context,
+	request *p.InternalCreateOrUpdateNexusIncomingServiceRequest,
+) error {
+	// TODO: setting up as a batch because soon we will need to maintain a service_name -> service_id mapping
```
[the remark this answers is not in the document]
Comment (bergundy) [reply common/persistence/cassandra/nexus_service_store.go:100]: You'll need to update the table version in this query.

File: buf.lock
[the remark this answers is not in the document]
Comment (pdoerner) [reply buf.lock:1]: Accidentally added it in my previous PR and I don't think it's actually necessary. Ditto the buf.yaml file

File: common/persistence/persistence_interface.go [hunk 4]
```diff
@@ -22,7 +22,7 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-//go:generate mockgen -copyright_file ../../LICENSE -package mock -source $GOFILE -destination mock/store_mock.go -aux_files go.temporal.io/server/common/persistence=data_interfaces.go
+//go:generate mockgen -copyright_file ../../LICENSE -package mock -source $GOFILE -destination mock/store_mock.go -aux_files go.temporal.io/server/common/incomingServiceStore=data_interfaces.go
```
[the remark this answers is not in the document]
Comment (pdoerner) [reply common/persistence/persistence_interface.go:25]: Didn't mean to, just a find and replace mistake. Thought I caught all of them.

File: common/persistence/cassandra/nexus_service_store.go [hunk 5]
```diff
@@ -61,30 +97,195 @@ func (s *NexusServiceStore) Close() {
...
+		request.LastKnownTableVersion)
+
+	previous := make(map[string]interface{})
+	applied, iter, err := s.session.MapExecuteBatchCAS(batch, previous)
+
+	if err != nil {
+		return gocql.ConvertError("CreateOrUpdateNexusIncomingService", err)
+	}
+
+	// We only care about the conflict in the first query
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:141]: Why?

Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:141]: I'm not sure I understand this comment. Both queries may have conflicts.

File: common/persistence/cassandra/nexus_service_store.go [hunk 6]
```diff
@@ -34,6 +36,40 @@ import (
 	"go.temporal.io/server/common/persistence/nosql/nosqlplugin/cassandra/gocql"
 )
 
+const (
+	UnknownTableVersion = 0
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:40]: I don't think we need this constant, just use 0 throughout the code IMHO.

File: common/persistence/cassandra/nexus_service_store.go [hunk 7]
```diff
@@ -61,30 +97,195 @@ func (s *NexusServiceStore) Close() {
...
+	}
+
+	currentTableVersion, err := s.getTableVersion(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	response.TableVersion = currentTableVersion
+
+	if request.LastKnownTableVersion != UnknownTableVersion && request.LastKnownTableVersion != currentTableVersion {
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:241]: My problem with this approach is that the table version is inconsistent with the list result.
It works if the last known table version is not 0, otherwise, the list result may reflect an older version of the table.

For the first page, you actually want to list all rows including the partition state row. For the other pages, what you have here works.

File: common/persistence/cassandra/nexus_service_store.go [hunk 8]
```diff
@@ -61,30 +97,195 @@ func (s *NexusServiceStore) Close() {
...
+	return response, nil
 }
 
 func (s *NexusServiceStore) DeleteNexusIncomingService(
 	ctx context.Context,
-	name string,
+	request *p.InternalDeleteNexusIncomingServiceRequest,
 ) error {
-	return serviceerror.NewUnimplemented("Cassandra DeleteNexusIncomingService is not implemented")
+	batch := s.session.NewBatch(gocql.UnloggedBatch).WithContext(ctx)
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:256]: I'd add a comment explaining why we use unlogged batch.

File: common/persistence/cassandra/nexus_service_store.go [hunk 9]
```diff
@@ -61,30 +97,195 @@ func (s *NexusServiceStore) Close() {
...
+			request.Service.Data.EncodingType.String(),
+			request.Service.Version+1,
+			constServicesPartition,
+			rowTypeIncomingNexusService,
+			request.Service.ServiceID,
+			request.Service.Version,
+		)
+	}
+
+	batch.Query(templateUpdateTableVersion,
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:127]: You need the same check here for `LastKnownTableVersion == 0` and issue an insert statement.

File: common/persistence/persistence_interface.go [hunk 10]
```diff
@@ -191,10 +191,10 @@ type (
 	NexusServiceStore interface {
 		Closeable
 		GetName() string
-		GetNexusIncomingService(ctx context.Context, name string) (*InternalGetNexusIncomingServiceResponse, error)
+		CreateOrUpdateNexusIncomingService(ctx context.Context, request *InternalCreateOrUpdateNexusIncomingServiceRequest) error
+		GetNexusIncomingService(ctx context.Context, serviceID string) (*InternalNexusIncomingService, error)
```
[the remark this answers is not in the document]
Comment (bergundy) [reply common/persistence/persistence_interface.go:195]: Please delete this API

**Review by bergundy — COMMENTED** [review]
Getting closer

File: common/persistence/cassandra/nexus_service_store.go (hunk 7 above)
Comment (pdoerner) [reply common/persistence/cassandra/nexus_service_store.go:241]: My concern is that I'm not sure we can guarantee that the table version row will be in the first page. I'm still pretty new to CQL though, so if we can be sure of this, then I can add the special handling for the first page.

Comment (pdoerner) [reply common/persistence/cassandra/nexus_service_store.go:241]: Also, I think as it is written now, we would only return an inconsistent list result if an incorrect table version was provided in the request and then between getting the results and getting the current table version, the table was updated so that now that originally incorrect table version is the current table version. For example: table version is 4, request comes in with version=5, get result page, table gets updated so version=5, then we check the current version and the condition passes. Otherwise if there was an update in between the two selects, this condition will fail and the caller will have to refetch the page with the new version.

Comment (bergundy) [reply common/persistence/cassandra/nexus_service_store.go:241]: I think you can guarantee that and worst case add an order by clause.

File: common/persistence/cassandra/nexus_service_store.go [hunk 11]
```diff
@@ -34,6 +35,48 @@ import (
...
+	// table templates
+	templateCreateTableVersion = `INSERT INTO nexus_incoming_services(partition, type, service_id, version) VALUES (?, ?, ?, ?) IF NOT EXISTS`
+	templateGetTableVersion    = `SELECT version FROM nexus_incoming_services WHERE partition = ? AND type = ? AND service_id = ?`
+	TemplateUpdateTableVersion = `UPDATE nexus_incoming_services SET version = ? WHERE partition = ? AND type = ? AND service_id = ? IF version = ?`
+
+	// incoming service templates
+	templateCreateIncomingServiceQuery         = `INSERT INTO nexus_incoming_services(partition, type, service_id, data, data_encoding, version) VALUES(?, ?, ?, ?, ?, ?) IF NOT EXISTS`
+	templateBaseListIncomingServicesQuery      = `SELECT service_id, data, data_encoding, version FROM nexus_incoming_services WHERE partition = ?`
+	TemplateListIncomingServicesQuery          = templateBaseListIncomingServicesQuery + ` AND type = ?`
+	TemplateListIncomingServicesFirstPageQuery = templateBaseListIncomingServicesQuery + ` ORDER BY type ASC`
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:60]:
```suggestion
	templateUpdateTableVersion = `UPDATE nexus_incoming_services SET version = ? WHERE partition = ? AND type = ? AND service_id = ? IF version = ?`

	// incoming service templates
	templateCreateIncomingServiceQuery         = `INSERT INTO nexus_incoming_services(partition, type, service_id, data, data_encoding, version) VALUES(?, ?, ?, ?, ?, ?) IF NOT EXISTS`
	templateBaseListIncomingServicesQuery      = `SELECT service_id, data, data_encoding, version FROM nexus_incoming_services WHERE partition = ?`
	templateListIncomingServicesQuery          = templateBaseListIncomingServicesQuery + ` AND type = ?`
	templateListIncomingServicesFirstPageQuery = templateBaseListIncomingServicesQuery + ` ORDER BY type ASC`
```

File: common/persistence/cassandra/nexus_service_store.go [hunk 12]
```diff
@@ -34,6 +35,48 @@ import (
...
+	templateUpdateIncomingServiceQuery         = `UPDATE nexus_incoming_services SET data = ?, data_encoding = ?, version = ? WHERE partition = ? AND type = ? AND service_id = ? IF version = ?`
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:61]: nit: put this next to the INSERT query

File: common/persistence/cassandra/nexus_service_store.go [hunk 13]
```diff
@@ -61,30 +104,282 @@ func (s *NexusServiceStore) Close() {
...
+				request.LastKnownTableVersion,
+				currentTableVersion)
+		}
+
+		currentServiceVersion, err := getTypedFieldFromRow[int64]("version", previousService)
+		if err != nil {
+			return fmt.Errorf("error retrieving current service version: %w", err)
+		}
+		if currentServiceVersion != request.Service.Version {
+			return fmt.Errorf("%w. Provided service version: %v Current service version: %v",
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:181]: Use lower case for Go errors.

```suggestion
			return fmt.Errorf("%w. provided service version: %v, current service version: %v",
```

File: common/persistence/cassandra/nexus_service_store.go [hunk 14]
```diff
@@ -61,30 +104,282 @@ func (s *NexusServiceStore) Close() {
...
+		return gocql.ConvertError("CreateOrUpdateNexusIncomingService", err)
+	}
+
+	if !applied {
+		currentTableVersion, err := getTypedFieldFromRow[int64]("version", previousPartitionStatus)
+		if err != nil {
+			return fmt.Errorf("error retrieving current table version: %w", err)
+		}
+		if currentTableVersion != request.LastKnownTableVersion {
+			return fmt.Errorf("%w. Provided table version: %v Current table version: %v",
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:170]:
```suggestion
			return fmt.Errorf("%w. provided table version: %v, current table version: %v",
```

File: common/persistence/cassandra/nexus_service_store.go [hunk 15]
```diff
@@ -61,30 +104,282 @@ func (s *NexusServiceStore) Close() {
...
+	return nil
 }
 
 func (s *NexusServiceStore) ListNexusIncomingServices(
 	ctx context.Context,
 	request *p.InternalListNexusIncomingServicesRequest,
 ) (*p.InternalListNexusIncomingServicesResponse, error) {
-	return nil, serviceerror.NewUnimplemented("Cassandra ListNexusIncomingServices is not implemented")
+	if request.PageSize <= 0 {
+		return nil, ErrNonPositiveListNexusIncomingServicesPageSize
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:200]: It's okay if you put this validation here but there should also be validation in the frontend that should return a bad request error (for the next PR).

File: common/persistence/cassandra/nexus_service_store.go (hunk 6 above)
Comment (bergundy) [reply common/persistence/cassandra/nexus_service_store.go:39]: Also no reason to export this AFAICT.

File: buf.lock
Comment (bergundy) [inline buf.lock:1]: This shouldn't be included in the PR.

File: buf.yaml
Comment (bergundy) [inline buf.yaml:1]: Same.

File: common/persistence/cassandra/nexus_service_store.go [hunk 16]
```diff
@@ -61,30 +104,282 @@ func (s *NexusServiceStore) Close() {
...
+		return gocql.ConvertError("DeleteNexusIncomingService", err)
+	}
+
+	if !applied {
+		currentTableVersion, err := getTypedFieldFromRow[int64]("version", previousPartitionStatus)
+		if err != nil {
+			return fmt.Errorf("error retrieving current table version: %w", err)
+		}
+		if currentTableVersion != request.LastKnownTableVersion {
+			return fmt.Errorf("%w. Provided table version: %v Current table version: %v",
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:280]: Same as above

File: common/persistence/cassandra/nexus_service_store.go [hunk 17]
```diff
@@ -61,30 +104,282 @@ func (s *NexusServiceStore) Close() {
...
+	currentTableVersion, err := s.getTableVersion(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	response.TableVersion = currentTableVersion
+
+	if request.LastKnownTableVersion != currentTableVersion {
+		// If table has been updated during pagination, throw error to indicate caller must start over
+		return nil, fmt.Errorf("%w. Provided table version: %v Current table version: %v",
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:235]: same as my comment above.

File: common/persistence/cassandra/nexus_service_store.go [hunk 18]
```diff
@@ -61,30 +104,282 @@ func (s *NexusServiceStore) Close() {
...
+		return nil, err
+	}
+	response.Services = services
+
+	if len(iter.PageState()) > 0 {
+		response.NextPageToken = iter.PageState()
+	}
+
+	if err := iter.Close(); err != nil {
+		return nil, serviceerror.NewUnavailable(fmt.Sprintf("ListNexusIncomingServices operation failed. Error: %v", err))
```
Comment (bergundy) [inline common/persistence/cassandra/nexus_service_store.go:223]:
```suggestion
		return nil, serviceerror.NewUnavailable(fmt.Sprintf("ListNexusIncomingServices operation failed: %v", err))
```

File: common/persistence/tests/cassandra_test.go [hunk 19]
```diff
@@ -1215,3 +1235,265 @@ func deleteMessages(
...
+	t.Run("ConcurrentCreateAndUpdate", func(t *testing.T) {
+		testCassandraNexusIncomingServiceStoreConcurrentCreateAndUpdate(t, store, tableVersion)
+	})
+	t.Run("ConcurrentUpdateAndDelete", func(t *testing.T) {
+		testCassandraNexusIncomingServiceStoreConcurrentUpdateAndDelete(t, store, tableVersion)
+	})
+}
+
+func testCassandraNexusIncomingServiceStoreConcurrentCreate(t *testing.T, store persistence.NexusServiceStore, tableVersion *atomic.Int64) {
+	ctx := context.Background()
```
Comment (bergundy) [inline common/persistence/tests/cassandra_test.go:1256]: Can we put a timeout on the test context?

**Review by bergundy — COMMENTED** [review]
Mostly nits, code looks great.

File: common/persistence/cassandra/nexus_service_store.go (hunk 6 above)
Comment (pdoerner) [reply common/persistence/cassandra/nexus_service_store.go:39]: My thinking is it will be needed at the manager level, but I will un-export it for now

File: common/persistence/tests/cassandra_test.go (hunk 19 above)
Comment (pdoerner) [reply common/persistence/tests/cassandra_test.go:1256]: I would rather not right now since we don't have it on any of the other tests AFAICT and they execute fairly quickly

Comment (bergundy) [reply common/persistence/tests/cassandra_test.go:1256]: If there's an issue with a test, the run may hang.
It's good practice IMHO.

**Review by bergundy — APPROVED** [review]
